### PR TITLE
DDP-5658 support revisioned templates

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

See DDP-5658. Database supports storing revisioned templates, but our queries are not accounting for revisioned "template variable substitutions". This PR refactors the code to support such revisioned templates. This will make it easier for studies to create new revisions of activities that requires only simple word change to templates, as in DDP-5583.

Key concepts:
* Templates contain a list of variables, and each variable have a list of translation substitution texts.
* These variable substitutions can be revisioned using time-based approach (e.g. start date, end date).
* So when querying for revisioned templates, we'll need a timestamp to pinpoint the revision.
* When querying for activity definitions, we use the activity's version start time as timestamp.
* When querying for activity instances, we use the instance's created_at time as timestamp.

I have ran several studies locally, and tested fetching of activity data for various revisions. It looks fine for me locally as far as I can tell. When this is merged, I'll ask for Kiara's help to thoroughly regression test this.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [x] :sweat_smile: There might be some issues here

## How do we demo these changes?

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivity.java [hunk 1]
```diff
@@ -180,10 +180,6 @@ int updateMaxInstancesPerUserById(
             @Bind("maxInstances") Integer maxInstances
     );
 
-    default List<ActivityValidationDto> findValidationsById(long activityId, long languageCodeId) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivity.java:184]: This method is not used, so I removed it.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivityValidation.java [hunk 2]
```diff
@@ -58,17 +56,6 @@ int _insertActivityValidation(
             @Bind("errorMessageTemplateId") long errorMessageTemplateId
     );
 
-    default List<ActivityValidationDto> _findByActivityIdTranslated(long activityId, long isoLanguageCode) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiActivityValidation.java:61]: This method is not used.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java [hunk 3]
```diff
@@ -370,7 +371,10 @@ public MailAddress verifyAddress(MailAddress address) throws AddressVerification
                             .findFirst()
                             .orElse(null);
                     if (warningTmplId != null) {
-                        msg = handle.attach(TemplateDao.class).loadTemplateById(warningTmplId).render(langCode);
+                        // Query using the current time to get the latest template.
+                        msg = handle.attach(TemplateDao.class)
+                                .loadTemplateByIdAndTimestamp(warningTmplId, Instant.now().toEpochMilli())
+                                .render(langCode);
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java:377]: NPE probability: loadTemplateByIdAndTimestamp can return null

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/StatisticsUtil.java [hunk 4]
```diff
@@ -84,7 +85,10 @@ private static StatisticsFigure generateStatisticsItem(Handle handle,
                 options.add(option);
                 optionByStableId.put(option.getStableId(), option);
             }
-            renderer.bulkRenderAndApply(handle, options, style, langId);
+
+            // As we see above, we're querying the latest question and picklist options,
+            // so we use the current time here to render the latest templates.
+            renderer.bulkRenderAndApply(handle, options, style, langId, Instant.now().toEpochMilli());
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/StatisticsUtil.java:91]: Does it make sense to create util method: `long epochMilli() {return Instant.now().toEpochMilli();}` ?
Then code could be a little bit compact.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg [hunk 5]
```diff
@@ -33,8 +41,12 @@ select t.template_id,
...
   left join i18n_template_substitution as ts on ts.template_variable_id = tv.template_variable_id
   left join language_code as lc on lc.language_code_id = ts.language_code_id
+  left join revision as ts_rev on ts_rev.revision_id = ts.revision_id
 >>
 
-queryTemplatesByIds(templateIds) ::= <<
-<select_all_templates()> where t.template_id in (<templateIds>)
+queryTemplatesByIdsAndTimestamp(templateIds) ::= <<
+<select_all_templates()>
+ where t.template_id in (<templateIds>)
```
Comment (evgeniipr) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:49]: Is it needed slashes before brackets: `<templateIds>` ?

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg [hunk 6]
```diff
@@ -8,11 +8,19 @@ from
...
     i18n_ts.language_code_id = :langCodeId
+  left join revision as i18n_ts_rev on
+    i18n_ts_rev.revision_id = i18n_ts.revision_id
   left join i18n_template_substitution as i18n_ts_def on
     i18n_ts_def.template_variable_id = tv.template_variable_id and
     i18n_ts_def.language_code_id = :defaultLangCodeId
+  left join revision as i18n_ts_def_rev on
+    i18n_ts_def_rev.revision_id = i18n_ts_def.revision_id
 where
   tv.template_id in (\<templateIds\>) and
```
Comment (evgeniipr) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:19]: I found out weird thing: slash before closing bracket seems not necessary: `\<templateIds>`.
I'm not 100% sure but when I tried to enable SQL logging for JDBI I encountered failures (fatal errors) with logging SQLs with params like this: in (\<templateIds\>). I experimented different ways and eventually made it working by removing the slash before closing bracket.
Btw, the logging can be activated in JDBI version 3.18.0 (in pepper now used 3.1.1) and it needs to add the following block of code in TransactionWrapper.java in constructor:
```
 SqlLogger sqlLogger = new SqlLogger() {
            @Override
            public void logAfterExecution(StatementContext context) {
                LOG.info("sql {}, parameters {}, timeTaken {} ms", context.getRenderedSql(),
                        context.getBinding().toString(), context.getElapsedTime(ChronoUnit.MILLIS));
            }
        };
        jdbi.setSqlLogger(sqlLogger);
```

**Review by evgeniipr — DISMISSED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/AddressService.java:377]: We checked that `warningTmplId != null` so we're expecting to get back an object here.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/StatisticsUtil.java (hunk 4 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/StatisticsUtil.java:91]: I think the way it is now isn't too bad.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg (hunk 5 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:49]: For this particular query, I found that we don't need slashes for escaping here.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg (hunk 6 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:19]: Yeah, whether we need slashes or not depends on a few things like, what the query is, if we're binding a list or not, and how the query is used in the code.

Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/TemplateDao.sql.stg:19]: Since this existing query has slashes, I'm keeping it as-is.

**Comment by yufengwng** [comment]
I found an issue with revisioned templates and activity-level validations while doing some local testing. I'm working on a fix.

**Review by evgeniipr — DISMISSED** [review]

**Comment by yufengwng** [comment]
@evgeniipr @MocanaAtBroad The issue is now fixed. See `BrainFixPrequalValidations` for more details. Basically, the Brain study configuration had some malformed revisions for the Prequal's activity validations. I added code to fix up those revision data.

**Review by evgeniipr — APPROVED** [review]

**Pull request merged**